### PR TITLE
optimize: eliminate the possibility of allocating too much direct memory

### PR DESCRIPTION
**Pull request by ggndnn**
### Ⅰ. Describe what this PR did
In FileTransactionStoreManager, sometimes, data buffer size was not under control. For example when data size to write was bigger than writerBuffer:

```
...
ByteBuffer byteBuffer = null;

if (bs.length + 4 > MAX_WRITE_BUFFER_SIZE) {
    //allocateNew
    byteBuffer = ByteBuffer.allocateDirect(bs.length + 4);
} else {
...
```
Or in function findTimeoutAndSave:
```
...
ByteBuffer byteBuffer = ByteBuffer.allocateDirect(totalSize);
for (byte[] bytes : listBytes) {
    byteBuffer.putInt(bytes.length);
    byteBuffer.put(bytes);
}
...
```
Those buffers' size was out of control, unlimited, that may be a hidden danger. 

This PR makes fully use of the writerBuffer, eliminate the possibility of allocating too much memory.

### Ⅲ. Why don't you add test cases (unit test/integration test)? 
Did.

### CODE REVIEW

File: server/src/main/java/io/seata/server/store/file/FileTransactionStoreManager.java [hunk 1]
```diff
@@ -215,46 +215,87 @@ private boolean saveHistory() throws IOException {
...
+        }
+        writeBuffer.putInt(dataLength);
+        bufferRemainingSize = writeBuffer.remaining();
+        int dataPos = 0;
+        while (dataPos < dataLength) {
+            int dataLengthToWrite = dataLength - dataPos;
+            dataLengthToWrite = Math.min(dataLengthToWrite, bufferRemainingSize);
+            writeBuffer.put(data, dataPos, dataLengthToWrite);
+            bufferRemainingSize = writeBuffer.remaining();
+            if (bufferRemainingSize <= 0) {
```
Comment (slievrly) [inline server/src/main/java/io/seata/server/store/file/FileTransactionStoreManager.java]: ==0

Comment (ggndnn) [reply server/src/main/java/io/seata/server/store/file/FileTransactionStoreManager.java]: Yes, it can't be < 0, just in case.

Comment (ggndnn) [reply server/src/main/java/io/seata/server/store/file/FileTransactionStoreManager.java]: Committed.

File: server/src/main/java/io/seata/server/store/file/FileTransactionStoreManager.java [hunk 2]
```diff
@@ -215,46 +215,87 @@ private boolean saveHistory() throws IOException {
...
-            LOGGER.error("save history data file error," + exx.getMessage());
+            LOGGER.error("save history data file error, {}", exx.getMessage(), exx);
             result = false;
         } finally {
             initFile(currFullFileName);
         }
         return result;
     }
 
+    private boolean writeDataFrame(byte[] data, int dataLength) {
```
Comment (jsbxyyx) [inline server/src/main/java/io/seata/server/store/file/FileTransactionStoreManager.java]: The dataLength passed in by the caller is data.length, whether to remove the dataLength parameter，use data.length??

Comment (ggndnn) [reply server/src/main/java/io/seata/server/store/file/FileTransactionStoreManager.java]: Function writeDataFrame is independent, has more functions than just specifying data as parameter, that could be more useful in future. I think no harm done to provide one more parameter dataLength. Do you agree?

Comment (jsbxyyx) [reply server/src/main/java/io/seata/server/store/file/FileTransactionStoreManager.java]: I am no problem, It’s just that developer are looking a little doubt by now.

Comment (ggndnn) [reply server/src/main/java/io/seata/server/store/file/FileTransactionStoreManager.java]: Ok, it's fine to remove dataLength parameter by now, clearer and more concise at least.

Comment (ggndnn) [reply server/src/main/java/io/seata/server/store/file/FileTransactionStoreManager.java]: Committed.

**Pull request merged**